### ISSUE TITLE
Fails silently when creating or destroying records

### ISSUE THREAD

**Issue by josacar** [chef-cookbooks/route53#60]
In `route53_record` when Amazon API gives a `400` error when `API Limit exceeded` or `Throttling`, error get swallowed by rescue clause:

``` ruby
  def delete
    zone(aws).records.get(name, type).destroy
    Chef::Log.debug("Destroyed record: #{name} #{type}")
  rescue Excon::Errors::BadRequest => e
    Chef::Log.error Nokogiri::XML(e.response.body).xpath('//xmlns:Message').text
  end
```

``` ruby
  def create
    begin
      zone(aws).records.create(record_attributes)
      Chef::Log.debug("Created record: #{record_attributes.inspect}")
    rescue Excon::Errors::BadRequest => e
      Chef::Log.error Nokogiri::XML( e.response.body ).xpath( "//xmlns:Message" ).text
    end
  end
```

I think this should fail and use the `retries` option in LWRPs or retry it with f.e. `retriable` gem. Failing silently is a bad idea

---

```
Excon::Errors::BadRequest
--------------------
Expected(200) <=> Actual(400 Bad Request)
excon.error.response
  :body          => "<?xml version=\"1.0\"?>\n<ErrorResponse xmlns=\"https://route53.amazonaws.com/doc/2013-04-01/\"><Error><Type>Sender</Type><Code>Throttling</Code><Message>Rate exceeded</Message></Error><RequestId>56cd4a47-8752-11a4-990a-2baaae76d214</RequestId></ErrorResponse>"
  :headers       => {
    "Content-Length"   => "255"
    "Content-Type"     => "text/xml"
    "Date"             => "Fri, 19 Dec 2014 07:40:43 GMT"
    "nnCoection"       => "close"
    "x-amzn-RequestId" => "56cd4a47-8752-11a4-990a-2baaae76d214"
  }
  :local_address => "XXXXXXXX"
  :local_port    => 41654
  :remote_ip     => "XXXXXXXX"
  :status        => 400
```

**Comment by josacar** [comment]
@slyness Any thoughts or suggestions on this?

**Comment by slyness** [comment]
@josacar I would agree that failing silently in this case is a bad idea. As long as it does not cause the entire convergence to fail as well. I would encourage you to submit a PR. It would be preferable to not include a separate gem requirement if possible.

**Comment by josacar** [comment]
There are several ways to attack this:

- Include a `excon` [middleware](https://github.com/mikehale/excon-middleware-aws-exponential_backoff)
- Use `idempotent` in excon [requests](https://github.com/fog/fog/issues/382) in fog
- Use a custom retry algorithm
- Raise error and let users handle it in Chef with `retries` option or just handling the error with custom code

**Comment by josacar** [comment]
@slyness Any preference?

**Comment by josacar** [comment]
There's another point where it can fail and it isn't handled in the provider, for example:

https://github.com/hw-cookbooks/route53/blob/develop/providers/record.rb#L69
https://github.com/hw-cookbooks/route53/blob/develop/providers/record.rb#L53
https://github.com/hw-cookbooks/route53/blob/develop/providers/record.rb#L55
https://github.com/hw-cookbooks/route53/blob/develop/providers/record.rb#L70

So IMO, third option makes no sense to me, neither leaving it as is.

**Comment by josacar** [comment]
@slyness `fog-aws` has the idempotency flag set in master https://github.com/fog/fog-aws/pull/40

**Comment by josacar** [comment]
@slyness `fog-aws` has the idempotency flag set in master https://github.com/fog/fog-aws/pull/40

**Comment by cheeseplus** [comment]
This should be fixed by virtue of the 1.0.0 rewrite to use the AWS SDK but please do open a new issue if the behaviour is still present in a current version.
